python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_direct_invocation.py

```diff
@@ -411,3 +411,50 @@ def an_asset(context, my_resource: NumResource, my_other_resource: NumResource)
     )
     assert executed["yes"]
     executed.clear()
+
+
+def test_direct_invocation_output_metadata():
+    @asset
+    def my_asset(context):
+        context.add_output_metadata({"foo": "bar"})
+
+    @asset
+    def my_other_asset(context):
+        context.add_output_metadata({"baz": "qux"})
+
+    ctx = build_op_context()
+
+    my_asset(ctx)
+    assert ctx.get_output_metadata("result") == {"foo": "bar"}
+
+    with pytest.raises(
+        DagsterInvariantViolationError,
+        match="attempted to log metadata for output 'result' more than once",
+    ):
+        my_other_asset(ctx)
+
+
+def test_async_assets_with_shared_context():
+    @asset
+    async def async_asset_one(context):
+        assert context.asset_key.to_user_string() == "async_asset_one"
+        await asyncio.sleep(0.01)
+        return "one"
+
+    @asset
+    async def async_asset_two(context):
+        assert context.asset_key.to_user_string() == "async_asset_two"
+        await asyncio.sleep(0.01)
+        return "two"
+
+    ctx = build_op_context()
```

Comment:
maybe leave a comment here with the context form our last discussion - that we could change this to require separate contexts per invocation but wanted to document that currently we support passing the same